### DISCUSSION TITLE
Programatically manipulate a Nickel document preserving comments

### DISCUSSION THREAD

**Discussion by wezm** [tweag/nickel#1779] [Q&A]
I'm interested in manipulating a Nickel document in a way that preserves comments. Is this possible?

**Use case/user story:**

Say I have a package specification like this:

```nickel
let pkg = {
  pkgname = "ripgrep",
  pkgver = "14.1.0",
  pkgrel = 0,
  build_style = "cargo",
  hostmakedepends = ["cargo"],
  makedepends = ["rust-std"],
  pkgdesc = "Tool that recursively searches the current directory for a regex pattern",
  maintainer = "Wesley Moore <wes@wezm.net>",
  license = "MIT OR Unlicense",
  url = "https://github.com/BurntSushi/ripgrep",
  source = [
    "%{url}/archive/%{pkgver}.tar.gz",
    # for pre-generated man pages
    "%{url}/releases/download/%{pkgver}/ripgrep-%{pkgver}-x86_64-unknown-linux-musl.tar.gz",
  ],
  source_paths = [".", "docs-prebuilt"],
  sha256 = [
    "33c6169596a6bbfdc81415910008f26e0809422fda2d849562637996553b2ab6",
    "f84757b07f425fe5cf11d87df6644691c644a5cd2348a2c670894272999d3ba7",
  ],
}
in

{
  package = pkg,
  hooks = {
    post_install = {
      license = ["LICENSE-MIT"],
      man = ["docs-prebuilt/doc/rg.1"],
      completion = [
        { file = "docs-prebuilt/complete/rg.bash", shell = 'bash, name = "rg" },
        { file = "docs-prebuilt/complete/rg.fish", shell = 'fish, name = "rg" },
        { file = "docs-prebuilt/complete/_rg", shell = 'zsh, name = "rg" },
      ]
    }
  }
}
```

A new version of ripgrep is released say 15.0.0 and I want to write a tool that can read the specification and update the `pkgver` and `sha256` fields. Something like `uppkg 15.0.0 ripgrep.ncl`.

Is this possible at the moment?

**Comment by yannham** [comment]
Unfortunately, this isn't possible currently, at least not in a trivial way. Interestingly I had to do a similar operation for Nickel's release script to `Cargo.toml` files, and even for TOML, I couldn't a common tool to do that - things like `tomlq` work by transpiling to and from `JSON`, but this obviously looses comments, structure, etc.

It might not help for your use-case, but note that if you want to override the version on-the-fly, you can use the (still experimental, but this won't probably change a lot in the future) customize mode of the CLI:

```console
$ nickel export pkg.ncl -- --override 'package.pkgver="15.0.0"'
[.. JSON export with version 15.0.0 ..]
```

But this doesn't change `pkg.ncl` itself, just override it on-the-fly.

Although the current parser loses some information, it still retain all expression positions, so I wonder if supporting an operation like that would be very difficult. Though I don't know right now what would be the interface - just a new subcommand to update a field in place? What if the value comes from a computation, or from merging?

**Reply by wezm** [reply]
Thanks for the reply. The overrides don't really work in this case because I intend for the Nickel files to be consumed via a Rust program and the version info needs to be statically known. While requiring two files isn't ideal the suggestion by jneem below might work for now.

**Comment by jneem** [comment]
There are only a few specific fields that you want to automatically update, right? What if you define them in a separate json file? That way you get a clear separation between the part that is automatically generated and the part that isn't:

```
let pkg_data = import "pkg-data.json" in
let pkg = {
  pkgname = "ripgrep",
  pkgver = pkg_data.pkgver,
  ...
}
```

**Reply by wezm** [reply]
Ahh good idea. Perhaps not ideal to need two files but this certainly solves the problem so I might proceed with it for now.
